File: nodejs-tools/nodejs-compiler-frontend/lexer-xml/src/readEscapeSequence.ts
/* eslint-disable @typescript-eslint/no-use-before-define */

import assert from 'assert';

import { EOF_CHARACTER, LexerError } from '@vighnesh153/lexer-core';
import { XmlLexer } from './Lexer';
import { HEXADECIMAL_DIGITS, repeat } from '@vighnesh153/utils';

export function readEscapeSequence(lexer: XmlLexer): string {
  assert.ok(
    lexer.inputReader.currentCharacter === '\\',
    `You should not attempt to read an escaped sequence if it doesn't start with '\\'`
  );

  // move over the backslash
  lexer.inputReader.readNextCharacter();

  const currCh: string = lexer.inputReader.currentCharacter;
  switch (currCh) {
    case 'u':
      return parseUnicode(lexer);
    case 't':
      return `\t`;
    case 'n':
      return `\n`;
    case '\\':
      return `\\`;
    case '"':
      return `"`;
    case EOF_CHARACTER: {
      lexer.addError(
        new LexerError({
          errorMessage: 'Unclosed escape sequence',
          lineNumber: lexer.inputReader.lineNumber,
          columnNumber: lexer.inputReader.columnNumber,
        })
      );
      return `\n`;
    }
    default: {
      lexer.addError(
        new LexerError({
          errorMessage: `Invalid escape character literal: ${lexer.inputReader.currentCharacter}`,
          lineNumber: lexer.inputReader.lineNumber,
          columnNumber: lexer.inputReader.columnNumber,
        })
      );
      return `\n`;
    }
  }
}

function parseUnicode(lexer: XmlLexer): string {
  assert.ok(
    lexer.inputReader.currentCharacter === 'u',
    `You should not try to parse a unicode sequence that doesn't begin with 'u'`
  );

  const unicodeCharacters: string[] = [];

  // unicode: \u0000 to \uFFFF
  repeat(4, () => {
    const peek = lexer.inputReader.peekCharacter()?.toLowerCase();
    if (HEXADECIMAL_DIGITS.includes(`${peek}`.toLowerCase())) {
      lexer.inputReader.readNextCharacter();
      unicodeCharacters.push(lexer.inputReader.currentCharacter!);
    } else {
      lexer.addError(
        new LexerError({
          errorMessage: `Invalid unicode character: ${peek}`,
          lineNumber: lexer.inputReader.lineNumber,
          columnNumber: lexer.inputReader.columnNumber,
        })
      );
      return ' ';
    }
  });

  return String.fromCharCode(parseInt(unicodeCharacters.join(''), 16));
}

File: jvm-tools/XmlFormatter/src/main/kotlin/dev/vighnesh153/xml/formatter/lexer/readEscapeSequence.kt
package dev.vighnesh153.xml.formatter.lexer

internal fun XmlLexer.readEscapeSequence(): Char? {
    assert(inputReader.currChar == BACKSLASH_CHARACTER) {
        "Attempting to read an escape sequence when start character isn't '\\'"
    }

    // move over the backslash
    inputReader.readNextChar()

    return when (inputReader.currChar) {
        'u' -> parseUnicodeCharacter()
        't' -> '\t'
        'n' -> '\n'
        '\\' -> '\\'
        '"' -> '"'
        EOF_CHARACTER -> {
            addError(
                UnclosedEscapeSequenceLexerError(
                    lineNumber = inputReader.lineNumber,
                    columnNumber = inputReader.columnNumber,
                )
            )
            null
        }

        else -> {
            addError(
                InvalidEscapeCharacterLexerError(
                    ch = inputReader.currChar,
                    lineNumber = inputReader.lineNumber,
                    columnNumber = inputReader.columnNumber,
                )
            )
            null
        }
    }
}

private fun XmlLexer.parseUnicodeCharacter(): Char? {
    assert(inputReader.currChar == 'u') {
        "Attempting to read a unicode character when start character isn't 'u'"
    }

    val unicodeCharacterBuilder = StringBuilder()

    // unicode: \u0000 to \uFFFF
    repeat(4) {
        val peek = inputReader.peekChar()?.lowercase()?.getOrNull(0)
        if (peek != null && (peek.isDigit() || peek in "abcdef")) {
            inputReader.readNextChar()
            unicodeCharacterBuilder.append(inputReader.currChar)
        } else {
            addError(
                InvalidUnicodeCharacterLexerError(
                    ch = peek,
                    lineNumber = inputReader.lineNumber,
                    columnNumber = inputReader.columnNumber,
                )
            )
            return null
        }
    }

    // https://stackoverflow.com/a/20239460/8822610
    return java.lang.String.valueOf(
        Character.toChars(
            Integer.parseInt(
                unicodeCharacterBuilder.toString(),
                16
            )
        )
    ).getOrNull(0)
}